src/canister_tests/src/certificate_validation.rs

```diff
@@ -0,0 +1,147 @@
+/// https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-gateway-certification
+pub fn validate_certification(
+    ic_certificate: &str,
+    canister_id: CanisterId,
+    uri_path: &str,
+    body: &[u8],
+    encoding: Option<&str>,
+    root_key: ThresholdSigPublicKey,
+) -> Result<(), ValidationError> {
+    let cert = parse_header_value(ic_certificate, 1)?;
+    let tree: HashTree = parse_header_value(ic_certificate, 2)?;
+    let agent = initialize_agent(&root_key);
+
+    if let Err(err) = agent.verify(&cert, canister_id.get().0, false) {
+        return Err(ValidationError::CertificateValidationFailed { inner: err });
+    }
+
+    let certified_data_path = vec![
+        "canister".into(),
+        canister_id.into(),
+        "certified_data".into(),
+    ];
+    let witness = match lookup_value(&cert, certified_data_path) {
+        Ok(witness) => witness,
+        Err(err) => {
+            return Err(WitnessLookupFailed { inner: err });
+        }
+    };
+
+    if *witness != tree.digest() {
+        return Err(TreeHashCertifiedDataMismatch);
+    }
+
+    let asset_path = ["http_assets".into(), uri_path.into()];
+    let tree_sha = match tree.lookup_path(&asset_path) {
+        LookupResult::Found(v) => v,
+        _ => match tree.lookup_path(&["http_assets".into(), "/index.html".into()]) {
+            LookupResult::Found(v) => v,
+            _ => {
+                return Err(AssetPathLookupFailed);
+            }
+        },
+    };
+
+    let body_sha = decode_body_to_sha256(body, encoding).unwrap();
+    if body_sha != tree_sha {
+        return Err(AssetHashMismatch);
+    }
+    Ok(())
+}
+
+fn initialize_agent(root_key: &ThresholdSigPublicKey) -> Agent {
+    let agent = Agent::builder()
+        .with_transport(
+            ReqwestHttpReplicaV2Transport::create("https://irrelevant-but-mandatory.com").unwrap(),
+        )
+        .build()
+        .unwrap();
+    agent
+        .set_root_key(public_key_to_der(&root_key.into_bytes()).unwrap())
+        .expect("setting root key failed");
+    agent
+}
+
+fn parse_header_value<T>(ic_certificate: &str, capture_group: usize) -> Result<T, ValidationError>
+where
+    T: for<'a> Deserialize<'a>,
+{
+    let captures = Regex::new(CERTIFICATE_REGEX)
+        .unwrap()
+        .captures(ic_certificate)
+        .ok_or(MalformedCertificate {
+            message: "unexpected format".to_string(),
+        })?;
+    let cert_blob = base64::decode(
+        captures
+            .get(capture_group)
+            .ok_or(MalformedCertificate {
+                message: format!("no regex match for capture group {}", capture_group),
+            })?
+            .as_str(),
+    )
+    .map_err(|err| MalformedCertificate {
+        message: format!("failed to decode base64 value: {:?}", err),
+    })?;
+    serde_cbor::from_slice(&cert_blob).map_err(|err| MalformedCertificate {
+        message: format!("failed to decode cbor value: {:?}", err),
+    })
```

Comment:
here it looks like you're parsing something like this:

`certificate=:<cert>:, tree=:<tree>:`

and returning either `cert` if `capture_group` is `1`, and `tree` if `capture_group` is `2` (that's what I gathered from a quick read).

Few questions:

* what happens if I pass `capture_group: 3`?
* do you ever call `parse_header_value` to use _only_ either `tree` or `cert`? i.e. does the caller really need to specify which one they need?

Reply:
Yeah, I restructured the code to be more clear (hopefully).